datasetinsights/io/tracker/factory.py

```diff
@@ -0,0 +1,115 @@
+        """create tracker class object.
+
+        Args:
+            config : config object, holds server details
+            tracker_type : type of tracker
+
+        Returns:
+            tracker instance.
+        """
+        if TrackerFactory.MLFLOW_TRACKER == tracker_type:
+
+            tracker = config.get(TrackerFactory.TRACKER, None)
+            if tracker and tracker.get(TrackerFactory.MLFLOW_TRACKER, None):
+                mlflow_config = tracker.get(TrackerFactory.MLFLOW_TRACKER)
+                host_id = mlflow_config.get(TrackerFactory.HOST_ID, None)
+                if host_id:
+                    client_id = mlflow_config.get(
+                        TrackerFactory.CLIENT_ID, None
+                    )
+                    exp_name = mlflow_config.get(TrackerFactory.EXP_NAME, None)
+                    TrackerFactory.update_run_name(mlflow_config)
+                    mlflow_tracker = TrackerFactory._mlflow_tracker_instance(
+                        host_id=host_id, client_id=client_id, exp_name=exp_name
+                    ).get_mlflow()
+                    mlflow_tracker.start_run(
+                        run_name=TrackerFactory.DEFAULT_RUN_NAME
+                    )
+                    return mlflow_tracker
+            return TrackerFactory._null_tracker()
+        else:
+            raise NotImplementedError(f"Unknown tracker {tracker_type}!")
+
+    @staticmethod
+    def _mlflow_tracker_instance(host_id=None, client_id=None, exp_name=None):
+
+        """Static instance access method.
+
+        Args:
+            host_id: MlTracker server host
+            client_id: MLFlow tracking server client id
+            exp_name: name of the experiment
+        Returns:
+            tracker singleton instance.
+        """
+        if not TrackerFactory.__tracker_instance:
+            with TrackerFactory.__singleton_lock:
+                if not TrackerFactory.__tracker_instance:
+                    TrackerFactory.__tracker_instance = MLFlowTracker(
+                        host_id=host_id, client_id=client_id, exp_name=exp_name
+                    )
+        return TrackerFactory.__tracker_instance
+
+    @staticmethod
+    def _null_tracker():
+        """private getter method to get singleton instance.
+
+        Returns:
+            NullTracker singleton instance.
+        """
+        if not TrackerFactory.__tracker_instance:
+            with TrackerFactory.__singleton_lock:
+                if not TrackerFactory.__tracker_instance:
+                    TrackerFactory.__tracker_instance = NullTracker()
+        return TrackerFactory.__tracker_instance
+
+    @staticmethod
+    def update_run_name(mlflow_config):
+
+        """Static method to update run name from config.
+
+        Args:
+            mlflow_config : mlflow_config dictionary
+        """
+        run_name = mlflow_config.get(TrackerFactory.RUN_NAME, None)
+        if run_name:
+            TrackerFactory.DEFAULT_RUN_NAME = run_name
```

Comment:
**????????????????**

- why does a static method in the factory take in params that look specific to mlflow?
- why override the default name?
- why isn't the name of the run updated through the tracker itself?

Reply:
why override the default name?
- If user pass run name from config then that should be the part of run else we need to disable the option to pass run name through the config.
why isn't the name of the run updated through the tracker itself?
why does a static method in the factory take in params that look specific to mlflow?
- Because I was starting run at factory and you need run name during start. 
- I will move this code to tracker and will start run there.